fence/man/fence_vmware.8
.TH fence_vmware 8

.SH NAME
fence_vmware - I/O Fencing agent for VMware virtual machines

.SH SYNOPSIS
.B 
fence_vmware
[\fIOPTION\fR]...

.SH DESCRIPTION
fence_vmware is an I/O Fencing agent which can be used with the VMware ESX or
VMware Server to fence virtual machines. It logs via ssh to a VMware ESX (or Server) 
physical machine and there run vmware-cmd, which does all work. 

By default, VMware ESX has disabled root account so you must create annother user 
account with limited permissions on VMWare ESX machine. 

Better idea is install right version of vmware-cmd (with same API) to any other machine 
in network and use it for fencing.

fence_vmware accepts options on the command line as well as from stdin.  
Fenced sends parameters through stdin when it execs the agent.  fence_vmware
can be run by itself with command line options.  This is useful for testing 
and for turning outlets on or off from scripts.

.SH OPTIONS
.TP
\fB-a\fP \fIIPaddress\fR
IP address or hostname of machine where is vmware-cmd.
.TP
\fB-A\fP \fIIPAdress\fR
IP adress or hostname of machine where is runing vmware. Default is localhost.
.TP
\fB-h\fP 
Print out a help message describing available options, then exit.
.TP
\fB-l\fP \fIlogin\fR
Login name of machine with vmware-cmd.
.TP
\fB-L\fP \fIlogin\fR
Login name to VMware administration interface.
.TP
\fB-o\fP \fIaction\fR
The action required.  This can be reboot (default), status, off or on.
.TP
\fB-p\fP \fIpassword\fR
Password for login or for passphrase for machine with vmware-cmd.
.TP
\fB-P\fP \fIpassword\fR
Password for login to VMware administration interface.
.TP
\fB-B\fP \fIscript\fR
Script to run to retrieve password for machine with vmware-cmd.
.TP
\fB-R\fP \fIscript\fR
Script to run to retrieve password for VMware administration interface.
.TP
\fB-n\fP \fIname\fR
Name of virtual machine to fence (ie. /vmfs/volumes/48bfcbd1-4624461c-8250-0015c5f3ef0f/Rhel/Rhel.vmx)
.TP
\fB-x\fP
Use secure connection over ssh (this is default, and can't be disabled) 
.TP
\fB-k\fP \fIfilename\fR
Identity file (private key) for ssh
.TP
\fB-T\fP
Test only.  Answer NO to the confirmation prompt instead of YES.
.TP
\fB-v\fP
Verbose. Record session to stdout, or debug file if specified (see -D).
.TP
\fB-D\fP
Specifies file, where will be written debug messages from session.
.TP
\fB-V\fP
Print out a version message, then exit.

.SH STDIN PARAMETERS
.TP
\fIagent = < param >\fR
This option is used by fence_node(8) and is ignored by fence_vmware.
.TP
\fIipaddr = < hostname | ip >\fR
IP address or hostname of machine where is vmware-cmd.
.TP
\fIvmipaddr = < hostname | ip >\fR
IP adress or hostname of machine where is runing vmware. Default is localhost.
.TP
\fIlogin = < param >\fR
Login name of machine with vmware-cmd.
.TP
\fIvmlogin = < param >\fR
Login name to VMware administration interface.
.TP
\fIoption = < param >\fR
The action required.  This can be reboot (default), status, off or on.
.TP
\fIpasswd = < param >\fR
Password for login or for passphrase for machine with vmware-cmd.
.TP
\fIvmpasswd = < param >\fR
Password for login to VMware administration interface.
.TP
\fIpasswd_script = < param >\fR
Script to run to retrieve password for machine with vmware-cmd.
.TP
\fIvmpasswd_script = < param >\fR
Script to run to retrieve password for VMware administration interface.
.TP
\fIport = < param >\fR
Name of virtual machine to fence (ie. /vmfs/volumes/48bfcbd1-4624461c-8250-0015c5f3ef0f/Rhel/Rhel.vmx)
.TP
\fIsecure = < param >\fR
Use secure connection over ssh (this is default, and can't be disabled) 
.TP 
\fIidentity_file = < param >\fR
Identity file (private key) for ssh
.TP
\fItest = < param >\fR
Test only.  Answer NO to the confirmation prompt instead of YES.
.TP
\fIverbose = < param >\fR
Verbose.  Record session to stdout, or debug file if specified (see debug).
.TP
\fIdebug = < param >\fR
Specifies file, where will be written debug messages from session.

.SH SEE ALSO
fence(8), fence_node(8)


fence/agents/vmware/fence_vmware.py
#!/usr/bin/python

#####
##
## The Following Agent Has Been Tested On VMware ESX 3.5 and VMware Server 1.0.7
## 
#####

import sys, re, pexpect
sys.path.append("@FENCEAGENTSLIBDIR@")
from fencing import *

#BEGIN_VERSION_GENERATION
RELEASE_VERSION="New VMware Agent - test release on steroids"
REDHAT_COPYRIGHT=""
BUILD_DATE=""
#END_VERSION_GENERATION

VMWARE_COMMAND="/usr/bin/vmware-cmd"
COMMAND_PROMPT_REG="\[PEXPECT\]\$ "
COMMAND_PROMPT_NEW="[PEXPECT]\$ "

# Start comunicating after login. Prepare good environment.
def start_communication(conn, options):
	conn.sendline ("PS1='"+COMMAND_PROMPT_NEW+"'");
	conn.log_expect(options,COMMAND_PROMPT_REG,SHELL_TIMEOUT)

# Prepare command line for vmware-cmd with parameters.
def prepare_cmdline(conn,options,add_vm_name):
	cmd_line=VMWARE_COMMAND+" -H '"+options["-A"]+"' -U '"+options["-L"]+"' -P '"+options["-P"]+"'"
	if (add_vm_name):
		cmd_line+=" '"+options["-n"]+"'"

        if options.has_key("-v"):
    		cmd_line+=" -v"

    	return cmd_line
    	
def get_power_status(conn, options):
	result = ""
	try:
		start_communication(conn,options)

		cmd_line=prepare_cmdline(conn,options,True)

            	cmd_line+=" getstate"

		conn.sendline(cmd_line)

		conn.log_expect(options,COMMAND_PROMPT_REG,SHELL_TIMEOUT)
		status_err = re.search("vmcontrol\ error\ ([-+]?\d+)\:(.*)",conn.before.lower())
		if (status_err!=None):
			fail_usage("VMware error "+status_err.group(1)+": "+status_err.group(2))
			
		status = re.search("getstate\(\)\ =\ on",conn.before.lower())
		
		result=(status==None and "off" or "on")
		
	except pexpect.EOF:
		fail(EC_CONNECTION_LOST)
	except pexpect.TIMEOUT:
		fail(EC_TIMED_OUT)

	return result

def get_outlet_list(conn,options):
	result={}

	try:
		start_communication(conn,options)

		cmd_line=prepare_cmdline(conn,options,False)
		cmd_line+=" -l"

		conn.sendline(cmd_line)

		conn.log_expect(options,COMMAND_PROMPT_REG,SHELL_TIMEOUT)
		status_err = re.search("vmcontrol\ error\ ([-+]?\d+)\:(.*)",conn.before.lower())
		if (status_err!=None):
			fail_usage("VMware error "+status_err.group(1)+": "+status_err.group(2))

                lines=conn.before.splitlines()

                for line in lines[(options.has_key("-v") and 3 or 1):-1]:
			if (line!=""):
			    result[line]=("","")

	except pexpect.EOF:
		fail(EC_CONNECTION_LOST)
	except pexpect.TIMEOUT:
		fail(EC_TIMED_OUT)

	return result

def set_power_status(conn, options):
	try:
		start_communication(conn,options)

		cmd_line=prepare_cmdline(conn,options,True)

            	cmd_line+=" "+(options["-o"]=="on" and "start" or "stop hard")

		conn.sendline(cmd_line)

		conn.log_expect(options,COMMAND_PROMPT_REG,POWER_TIMEOUT)

	except pexpect.EOF:
		fail(EC_CONNECTION_LOST)
	except pexpect.TIMEOUT:
		fail(EC_TIMED_OUT)
		
def main():
	device_opt = [  "help", "version", "agent", "quiet", "verbose", "debug",
			"action", "ipaddr", "login", "passwd", "passwd_script",
			"secure",  "identity_file", "test" , "vmipaddr", "vmlogin", 
			"vmpasswd", "port", "vmpasswd_script", "separator" ]

	options = check_input(device_opt, process_input(device_opt))

	## 
	## Fence agent specific defaults
	#####
        if 0 == options.has_key("-c"):
            options["-c"] = "\$ "

        if 0 == options.has_key("-A"):
            options["-A"] = "localhost"
    
    	options["-x"] = 1
	##
	## Operate the fencing device
	####
	conn = fence_login(options)
	fence_action(conn, options, set_power_status, get_power_status, get_outlet_list)

	##
	## Logout from system
	######
	conn.sendline("logout")
	conn.close()

if __name__ == "__main__":
	main()
